Optimum-Habana docs re-org

dsocek opened the pull request:
# Hugging Face Gaudi Documentation Update

Comprehensive update to HF Gaudi docs

libinta commented on docs/source/_toctree.yml:
```diff
@@ -32,20 +34,8 @@
     - local: usage_guides/multi_node_training
       title: Multi-node Training
+    - local: usage_guides/quantization
+      title: Quantization
```
why do we change this way?

dsocek replied:
see answer below

libinta commented on docs/source/index.mdx:
```diff
@@ -123,13 +122,5 @@ Besides, [this page](https://github.com/huggingface/optimum-habana/tree/main/exa
-    <a class="!no-underline border dark:border-gray-700 p-5 rounded-lg shadow hover:shadow-lg" href="./package_reference/trainer"
-      ><div class="w-full text-center bg-gradient-to-br from-purple-400 to-purple-500 rounded-lg py-1.5 font-semibold mb-5 text-white text-lg leading-relaxed">Reference</div>
-      <p class="text-gray-700">Technical descriptions of how the Habana classes and methods of 🤗 Optimum Habana work.</p>
-    </a>
```
why this is removed?

dsocek replied:
@libinta , please note this new, simplified organization was suggested by Greg.  As for why, I believe the intension was to simplify and reduce number of sections

libinta commented on docs/source/package_reference/distributed_runner.mdx:
```diff
@@ -1,20 +0,0 @@
-# DistributedRunner
-
-[[autodoc]] distributed.distributed_runner.DistributedRunner
-    - all
```
why removed this?

dsocek replied:
See Greg's vision plan below

dsocek replied:
Added back :)

libinta commented on docs/source/package_reference/stable_diffusion_pipeline.mdx:
```diff
@@ -1,85 +0,0 @@
-The `GaudiDDPMPipeline` is to enable unconditional image generations on HPUs. It has similar APIs as the regular `DiffusionPipeline`.
-It shares a common parent class, `GaudiDiffusionPipeline`, with other existing Gaudi pipelines. It now supports both DDPM and DDIM scheduler. 
-It is recommended to use the optimized scheduler, `GaudiDDIMScheduler`, to obtain the best performance and image outputs.
-
```
why remove this?

dsocek replied:
Greg suggested to upgrade to SDXL as SD is kind of old model now.  We can keep too, it may be a good intro diffuser model.

dsocek replied:
@kplau1128 can you comment here?

kplau1128 replied:
Greq was request and suggested consolidated SD into one tutorials/stable_diffusion.mdx, and remove others

libinta commented on docs/source/package_reference/trainer.mdx:
```diff
@@ -1,69 +0,0 @@
-## GaudiSeq2SeqTrainingArguments
-
-[[autodoc]] transformers.training_args_seq2seq.GaudiSeq2SeqTrainingArguments
-    - all
```
why removed?

dsocek replied:
Greg's vision of new doc org:
![image](https://github.com/user-attachments/assets/db8d47b7-31e6-431c-b2ce-5b39bebc059b)

dsocek replied:
Added back after yours and regis' review

libinta commented on docs/source/quickstart.mdx:
````diff
@@ -17,38 +17,52 @@ limitations under the License.
 ```
 
-where `gaudi_config_name` is the name of a model from the [Hub](https://huggingface.co/Habana) (Gaudi configurations are stored in model repositories) or a path to a local Gaudi configuration file (you can see [here](./package_reference/gaudi_config) how to write your own).
+where `gaudi_config_name` is the name of a model from the [Hub](https://huggingface.co/Habana) (Gaudi configurations are stored in model repositories)
````
why do we talk about gaudi_config_name here? it seems we are talking about docker before.

dsocek replied:
This out of place.  Good catch.  Probably leftover from previous content.

dsocek replied:
Good catch.  This is out of place, probably leftover from previous content

dsocek replied:
@libinta addressed suggested changes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_1488). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss reviewed (commented):
Thanks for updating the docs! Not agreeing with everything, but feel free to reply under my comments.

I would also keep the registration marks ® for titles only as it overloads the doc quite a lot and makes it look more like a marketing document. This is what we did for the README of Optimum Habana: https://github.com/huggingface/optimum-habana

dsocek commented on docs/source/tutorials/stable_diffusion_ldm3d.mdx:
Diffusing 3D latents topic is nice technological extension to SD, but seems a bit cherrypicked at end of 2024 when we have so many new extensions available in addition to base SD.  Also seems all new up-coming models are transformer based (DiTs instead of UNets).  IMO, we either keep SD section lean and clean or add bits on other extensions and derivatives to cover the field (SDXL, SD3, control nets, SVD).

dsocek commented on docs/source/concept_guides/hpu.mdx:
Keep but move out of removed `concept_guides` section

dsocek replied:
Added a basics on HPU/Gaudi section in index instead to consolidate, please check if this suffices

dsocek commented on docs/source/usage_guides/model_conversion.mdx:
Agreed this can be improved and moved

dsocek replied:
Replaced `model conversion` with `adaptation`, hopefully more clear as to we "adapt" the generic HF scripts/pipelines to HPU with small changes

dsocek commented on docs/source/package_reference/distributed_runner.mdx:
Greg's vision was to remove; however, we can also keep these IMO as well.  I can revert them back

dsocek replied:
reverted the section back

dsocek commented on docs/source/tutorials/gaudi_config.mdx:
The idea was to minimize number of sections.   Can you think of another better place for gaudi config file write-up?

dsocek replied:
Actually OK, if we revert `package_reference` section then it nicely goes back there

dsocek replied:
moved back to `package_reference` section

dsocek commented on docs/source/tutorials/single_hpu.mdx:
will revert it back, not sure why/how got removed

dsocek replied:
Reverted

dsocek commented on docs/source/tutorials/inference.mdx:
```diff
@@ -16,7 +16,10 @@ limitations under the License.
 
-This section shows how to run inference-only workloads on Gaudi.
+In most cases, you should start with the docuemeted examples in the optimum for Intel® Gaudi® library, as these arlready use the GaudiTrainer class.
+These are the best place to start running inference on the Intel® Gaudi® Accelerator
```
Removed this sentence, and added a sentence which links to examples in OH instead

dsocek commented on docs/source/tutorials/inference.mdx:
```diff
@@ -16,7 +16,10 @@ limitations under the License.
 # Run Inference
 
-This section shows how to run inference-only workloads on Gaudi.
+In most cases, you should start with the docuemeted examples in the optimum for Intel® Gaudi® library, as these arlready use the GaudiTrainer class.
```
Added link

dsocek commented on docs/source/index.mdx:
```diff
@@ -15,15 +15,23 @@ limitations under the License.
+Optimum for Intel Gaudi AI accelerator is the interface between the Transformers and Diffusers libraries and [Intel Gaudi AI Accelerators (HPUs)](https://docs.habana.ai/en/latest/index.html).
 It provides a set of tools that enable easy model loading, training and inference on single- and multi-HPU settings for various downstream tasks as shown in the table below.
 
-HPUs offer fast model training and inference as well as a great price-performance ratio.
-Check out [this blog post about BERT pre-training](https://huggingface.co/blog/pretraining-bert) and [this post benchmarking Intel Gaudi 2 with NVIDIA A100 GPUs](https://huggingface.co/blog/habana-gaudi-2-benchmark) for concrete examples.
-If you are not familiar with HPUs, we recommend you take a look at [our conceptual guide](./concept_guides/hpu).
```
Added blogs to index page

dsocek commented on docs/source/quickstart.mdx:
```diff
@@ -17,43 +17,53 @@ limitations under the License.
+🤗 Optimum for Intel Gaudi also features HPU-optimized support for the 🤗 Diffusers library.
 Thus, you can easily deploy Stable Diffusion on Gaudi for performing text-to-image generation.
 
 Here is how to use it and the differences with the 🤗 Diffusers library:
```
Done

dsocek commented on docs/source/quickstart.mdx:
Added small section on training

dsocek commented:
@regisss your comments have been (mostly) addressed.  Let me know if I missed anything. Doc can be a bit more subjective than code, cool about disagreements.  As I mentioned, Greg's vision was basis for this PR (it tries to follow "less is more principle" thus number of sections simplification)

auvarovahabana commented on docs/source/index.mdx:
```diff
@@ -109,27 +126,9 @@ In the tables below, ✅ means single-card, multi-card and DeepSpeed have all be
-    </a>
-  </div>
-</div>
+For some concrete examples, check out the following blogs:
```
For specific examples, check out the following blogs:

dsocek replied:
Done

auvarovahabana commented on docs/source/index.mdx:
```diff
@@ -15,15 +15,32 @@ limitations under the License.
+[Intel Gaudi 1](https://habana.ai/training/gaudi),
+[Intel Gaudi 2](https://habana.ai/training/gaudi2), and
+[Intel Gaudi 3](https://habana.ai/training/gaudi3).
+Each server is equipped with 8 devices, known as Habana Processing Units (HPUs), offering 32GB/96GB/128GB of memory on Gaudi 1/2/3.
```
on Gaudi 3, Gaudi 2, and first-gen Gaudi.

dsocek replied:
Done

auvarovahabana commented on docs/source/quickstart.mdx:
```diff
@@ -17,51 +17,152 @@ limitations under the License.
-- from transformers import Trainer, TrainingArguments
-+ from optimum.habana import GaudiTrainer, GaudiTrainingArguments
+## Accessing Intel Gaudi AI Accelerator
+To access an Intel Gaudi AI accelerator node in the Intel® Tiber® Developer cloud, you will go to
```
Intel® Tiber™ AI Cloud

dsocek replied:
Done

auvarovahabana commented on docs/source/quickstart.mdx:
```diff
@@ -17,51 +17,152 @@ limitations under the License.
-## Stable Diffusion
+</Tip>
+
+## Multi Device Inference
```
Multi-device

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/accelerate_inference.mdx:
```diff
@@ -100,3 +110,15 @@ outputs = generator(
+
+## Custom Operators
+
+Habana provides a few custom operators that achieve better performance than their PyTorch counterparts on Gaudi.
```
Intel Gaudi provides...

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/deepspeed.mdx:
```diff
@@ -28,16 +29,16 @@ You can find more information about DeepSpeed Gaudi integration [here](https://d
 ## Setup
 
-To use DeepSpeed on Gaudi, you need to install Optimum Habana and [Habana's DeepSpeed fork](https://github.com/HabanaAI/DeepSpeed) with:
+To use DeepSpeed on Gaudi, you need to install Optimum for Intel Gaudi and [Habana's DeepSpeed fork](https://github.com/HabanaAI/DeepSpeed) with:
```
To use DeepSpeed on Gaudi, you need to install Optimum for Intel Gaudi and [DeepSpeed fork]

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/multi_node_training.mdx:
```diff
@@ -26,70 +26,23 @@ Using several Gaudi servers to perform multi-node training can be done easily. T
 Two types of configurations are possible:
 - scale-out using Gaudi NICs or Host NICs (on-premises)
-- scale-out using AWS DL1 instances
+- scale-out using Intel Developer Cloud (IDC) instances
```
scale-out using Intel® Tiber™ AI Cloud instances

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/multi_node_training.mdx:
```diff
@@ -26,70 +26,23 @@ Using several Gaudi servers to perform multi-node training can be done easily. T
 
 
-### AWS DL1 instances
+### Intel Developer Cloud (IDC) instances
```
Intel Tiber AI Cloud

auvarovahabana commented on docs/source/usage_guides/multi_node_training.mdx:
```diff
@@ -123,15 +76,13 @@ env_variable_2_name=value
 - Larger batch sizes should lead to higher speedups.
 - Multi-node inference is not recommended and can provide inconsistent results.
-- On AWS DL1 instances, run your Docker containers with the `--privileged` flag so that EFA devices are visible.
+- On Intel Developer Cloud (IDC) instances, run your Docker containers with the `--privileged` flag so that EFA devices are visible.
```
Intel Tiber AI Cloud

dsocek replied:
Done

auvarovahabana commented on docs/source/quickstart.mdx:
```diff
@@ -17,51 +17,152 @@ limitations under the License.
-+ from optimum.habana import GaudiTrainer, GaudiTrainingArguments
+## Accessing Intel Gaudi AI Accelerator
+To access an Intel Gaudi AI accelerator node in the Intel® Tiber® Developer cloud, you will go to
+[Intel Developer Cloud Console](https://console.cloud.intel.com/hardware) and access the hardware instances to select the Intel Gaudi AI accelerator
```
Intel Tiber AI Cloud

dsocek replied:
Done

auvarovahabana commented on docs/source/package_reference/gaudi_config.mdx:
```diff
@@ -16,20 +16,109 @@ limitations under the License.
+Parameter values of this class can be set from an external JSON file.
 
-You can find examples of Gaudi configurations in the [Habana model repository on the Hugging Face Hub](https://huggingface.co/habana). For instance, [for BERT Large we have](https://huggingface.co/Habana/bert-large-uncased-whole-word-masking/blob/main/gaudi_config.json):
+You can find examples of Gaudi configurations in the [Habana model repository on the Hugging Face Hub](https://huggingface.co/habana).
```
Intel Gaudi model repository

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -77,10 +86,10 @@ The flexibility of PyTorch comes at a price - usually the same pythonic logic is
 This may lead to a situation where it takes longer for CPU to schedule the work on Habana accelerator than it is effectively computed by it.
```
This may lead to a situation where it takes longer for CPU to schedule the work on Gaudi than it is effectively computed by it.

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -165,6 +174,10 @@ The default value of *epsilon* is `1e-6` for the Habana fused ADAM optimizer, wh
 Habana provides a [custom gradient norm clipping implementation](https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Custom_Ops_PyTorch.html#other-custom-ops).
```
Intel Gaudi provides...
Change everywhere

dsocek replied:
Done

auvarovahabana commented on docs/source/usage_guides/deepspeed.mdx:
```diff
@@ -20,6 +20,7 @@ limitations under the License.
 - **ZeRO-2**: partitions the optimizer states + gradients across processes.
+- **ZeRO-3**: ZeRO-2 + full model state is partitioned across the processes.
 
 These configurations are fully compatible with Habana Mixed Precision and can thus be used to train your model in *bf16* precision.
```
Intel Gaudi Mixed Precision

dsocek replied:
Done

dsocek commented:
@auvarovahabana added all of your suggested changes/corrections. I also squashed last commit with original commit, and added all who chimed in as co-authors of this doc update.  @regisss and @libinta could you please do a final review/check ?

dsocek commented:
@regisss fixed coding style (although the fixed files had nothing to do with this PR ;)

regisss commented:
> @regisss fixed coding style (although the fixed files had nothing to do with this PR ;)

Yeah it was due to a new release of `ruff` :grin: 
Going to do another review in the next few hours

regisss commented on docs/source/index.mdx:
```diff
@@ -109,27 +126,9 @@ In the tables below, ✅ means single-card, multi-card and DeepSpeed have all be
-    </a>
-  </div>
-</div>
+For specific examples, check out the following blogs:
+- [Benchmarking Intel Gaudi 2 with NVIDIA A100 GPUs](https://huggingface.co/blog/habana-gaudi-2-benchmark)
+- [Accelerating Vision-Language Models: BridgeTower on Habana Gaudi2](https://huggingface.co/blog/bridgetower)
```
I would rather have these blog posts mentioned before the table of validated models because they hardly get any visibility here

dsocek replied:
Done

regisss commented on docs/source/tutorials/inference.mdx:
```diff
@@ -16,7 +16,16 @@ limitations under the License.
+An effective quick start would be to review the inference examples provided in the Optimum for Intel Gaudi
+[here].
+
+An effective quick start is to explore the 
```
```suggestion
You can also explore the 
```
To avoid using the exact same words as the previous paragraph?

regisss commented on docs/source/tutorials/stable_diffusion.mdx:
````diff
@@ -56,7 +56,7 @@ outputs = pipeline(
 ```
 
-Outputs can be PIL images or Numpy arrays. See [here](https://huggingface.co/docs/optimum/habana/package_reference/stable_diffusion_pipeline#optimum.habana.diffusers.StableDiffusionPipeline.__call__) all the parameters you can set to tailor generations to your taste.
+Outputs can be PIL images or Numpy arrays. See `StableDiffusionPipeline` all the parameters you can set to tailor generations to your taste.
````
In the second sentence, either we keep the link either we rephrase it but the current version doesn't mean anything IMO.

dsocek replied:
Done

regisss commented on docs/source/usage_guides/model_adaptation.mdx:
```diff
@@ -0,0 +1,83 @@
+-->
+
+
+# Model Adaptation
```
Still not convinced by this naming. "GPU migration" sounds good to me as used in the Gaudi doc here: https://docs.habana.ai/en/latest/PyTorch/PyTorch_Model_Porting/GPU_Migration_Toolkit/GPU_Migration_Toolkit.html
WDYT?

dsocek replied:
> Still not convinced by this naming. "GPU migration" sounds good to me as used in the Gaudi doc here: https://docs.habana.ai/en/latest/PyTorch/PyTorch_Model_Porting/GPU_Migration_Toolkit/GPU_Migration_Toolkit.html WDYT?

This one is a but sticky.   I think that Model Migration refers to process of performing Migration from general purpose GPU to Gaudi HPU.   However, in OH, we kind of migrated this by providing Gaudi specific pipeline.  So the user kind of needs to just "adapt" a code from HF libs (diffusers, transformers, ..) to OH in that sense.  Like instead of StableDiffuseion pipeline adapt and use GaudiStableDiffusion (didn't actually perform model migration).   It is my understanding but maybe I am totally off here :)   Your thoughts?

regisss replied:
The way I see it, this page shows how to port a Transformers/Diffusers script from GPU to HPU.
IMO "model adaptation" could mean many different things and users don't really have to know how different models are to use them on Gaudi. That's why any title with "model" sounds a bit odd to me.
Maybe this could be called "Adapt a Transformers/Diffusers script to Intel Gaudi". This is a bit verbose but at least much clearer regarding the content of this page.
But anyway let's merge this PR today and we can modify this later if we don't find something good here.

dsocek replied:
Ok I think you nailed the issue.  Title does come across as confusing and could be misleading as different people may understand it differently.  I will change it.

The actual Model Migration page would be actually describing steps for model migration, like taking existing say transformers or diffusers pipeline and migrating to Gaudi.  Maybe something along these lines can be added in future (for more advanced, daredevil developers :)

dsocek replied:
Fixed the title just now.  Also fixed the overview to better convey the pages organization in `usage_guides` section

regisss commented on docs/source/usage_guides/overview.mdx:
```diff
@@ -16,10 +16,12 @@ limitations under the License.
-- [Pretraining models](./pretraining)
-- [Accelerating training](./accelerate_training)
-- [Accelerating inference](./accelerate_inference)
+- [Model conversion](./model_conversion) show how to do model conversion
```
Please modify the title here too

dsocek replied:
Done

regisss commented on docs/source/quickstart.mdx:
````diff
@@ -83,23 +184,105 @@ outputs = pipeline(
 )
 ```
 
+## Training on Gaudi
````
Maybe move training before the diffusion section?

dsocek replied:
Done

regisss reviewed (commented):
I left a few additional comments, we should be able to merge this one tomorrow :slightly_smiling_face:

regisss commented on docs/source/index.mdx:
```diff
@@ -15,15 +15,36 @@ limitations under the License.
+The Intel Gaudi AI accelerator family currently includes three product generations:
+[Intel Gaudi 1](https://habana.ai/training/gaudi),
+[Intel Gaudi 2](https://habana.ai/training/gaudi2), and
+[Intel Gaudi 3](https://habana.ai/training/gaudi3).
```
```suggestion
[Intel Gaudi 3](https://habana.ai/products/gaudi3/).
```
https://habana.ai/training/gaudi3 doesn't exist anymore it seems. Actually even https://habana.ai/training/gaudi and https://habana.ai/training/gaudi2 redirects to https://habana.ai/products/gaudi and https://habana.ai/products/gaudi2.

dsocek replied:
Wow this works few days ago!  Ill update now

dsocek replied:
Fixed

regisss commented on docs/source/quickstart.mdx:
```diff
@@ -83,23 +239,50 @@ outputs = pipeline(
+  - [stable-video-diffusion](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion)
+  - [video-classification](https://github.com/huggingface/optimum-habana/tree/main/examples/video-classification)
+
+To learn more about how to migrate your own model, check out [Model Conversion](https://huggingface.co/docs/optimum/habana/usage_guides/model_conversion).
```
We need to update this link since we changed the name "Model conversion"

dsocek replied:
Fixed

regisss reviewed (commented):
I spotted a couple of dead links to update, and then I'll merge!
